src/graph/sampling/metapath_randomwalk.h

```diff
@@ -0,0 +1,147 @@
+  // uniform case), we want to reduce the overheads (even from object copies or object
+  // construction) as much as possible.
+  // Using Successors() slows down by 2x.
+  // Using OutEdges() slows down by 10x.
+  const auto &csr_arrays = edges_by_type[etype];
+  const IdxType *offsets = static_cast<IdxType *>(csr_arrays[0]->data);
+  const IdxType *all_succ = static_cast<IdxType *>(csr_arrays[1]->data);
+  const IdxType *succ = all_succ + offsets[curr];
+
+  int64_t size = offsets[curr + 1] - offsets[curr];
+  if (size == 0)
+    return std::make_pair(-1, true);
+
+  FloatArray prob_etype = prob[etype];
+  IdxType idx;
+  if (prob_etype->shape[0] == 0) {
+    // empty probability array; assume uniform
+    idx = RandomEngine::ThreadLocal()->RandInt(size);
+  } else {
+    // non-uniform random walk
+    const IdxType *all_eids = static_cast<IdxType *>(csr_arrays[2]->data);
+    const IdxType *eids = all_eids + offsets[curr];
+
+    ATEN_FLOAT_TYPE_SWITCH(prob_etype->dtype, DType, "probability", {
+      FloatArray prob_selected = FloatArray::Empty({size}, prob_etype->dtype, prob_etype->ctx);
+      DType *prob_selected_data = static_cast<DType *>(prob_selected->data);
+      const DType *prob_etype_data = static_cast<DType *>(prob_etype->data);
+      for (int64_t j = 0; j < size; ++j)
+        prob_selected_data[j] = prob_etype_data[eids[j]];
+      idx = RandomEngine::ThreadLocal()->Choice<IdxType>(prob_selected);
+    });
+  }
+  curr = succ[idx];
+
+  return std::make_pair(curr, terminate(data, curr, len));
+}
+
+/*!
+ * \brief Metapath-based random walk.
+ * \param hg The heterograph.
+ * \param seeds A 1D array of seed nodes, with the type the source type of the first
+ *        edge type in the metapath.
+ * \param metapath A 1D array of edge types representing the metapath.
+ * \param prob A vector of 1D float arrays, indicating the transition probability of
+ *        each edge by edge type.  An empty float array assumes uniform transition.
+ * \param terminate Predicate for terminating a random walk path.
+ * \return A 2D array of shape (len(seeds), len(metapath) + 1) with node IDs.
+ */
+template<DLDeviceType XPU, typename IdxType>
+IdArray MetapathBasedRandomWalk(
+    const HeteroGraphPtr hg,
+    const IdArray seeds,
+    const TypeArray metapath,
+    const std::vector<FloatArray> &prob,
+    TerminatePredicate terminate) {
+  int64_t max_num_steps = metapath->shape[0];
+  const IdxType *metapath_data = static_cast<IdxType *>(metapath->data);
+
+  // Prefetch all edges.
+  // This forces the heterograph to materialize all OutCSR's before the OpenMP loop;
+  // otherwise data races will happen.
+  // TODO(BarclayII): should we later on materialize COO/CSR/CSC anyway unless told otherwise?
+  std::vector<std::vector<IdArray> > edges_by_type;
+  for (dgl_type_t etype = 0; etype < hg->NumEdgeTypes(); ++etype)
+    edges_by_type.push_back(hg->GetAdj(etype, true, "csr"));
```

Comment:
Will you fix this in PR #1224 ?

Reply:
Materializing the CSR arrays beforehand is much faster.